File: app/seo.py
"""
SEO utilities for all pages. Provides professional meta titles, descriptions,
keywords, and Open Graph data for consistent, crawl-friendly content.
"""
import re
from typing import Optional, Any, List
from dataclasses import dataclass

try:
    from app.config.settings import RSS_FEEDS_BY_CATEGORY
except ImportError:
    RSS_FEEDS_BY_CATEGORY = {}

# Site branding (can be overridden via env if needed)
SITE_NAME = "Sugoi News"
SITE_TAGLINE = "Học tiếng Nhật qua tin tức - Tổng hợp & dịch bằng AI"
DEFAULT_OG_IMAGE = "/static/img/og-default.png"  # optional fallback
META_DESC_MAX_LEN = 160
KEYWORDS_HOME = (
    "học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, tin tức Nhật Bản tiếng Việt, tổng hợp tin tức Nhật Bản, Mainichi dịch tiếng Việt, "
    "học tiếng Nhật qua đọc tin tức, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, học từ vựng tiếng Nhật qua tin tức, đọc hiểu tiếng Nhật qua bài báo"
)

# All category names for "all categories" SEO (from config)
def _all_category_names() -> List[str]:
    return list(RSS_FEEDS_BY_CATEGORY.keys()) if RSS_FEEDS_BY_CATEGORY else []


def _keywords_all_categories() -> str:
    cats = _all_category_names()
    if not cats:
        return KEYWORDS_HOME
    return ", ".join(cats) + ", " + KEYWORDS_HOME


@dataclass
class SEOData:
    """SEO data for a page: title, description, keywords, image, canonical, etc."""
    title: str
    description: str
    keywords: str
    image: Optional[str] = None
    canonical_url: Optional[str] = None
    og_type: str = "website"
    article_published_time: Optional[str] = None
    article_author: Optional[str] = None
    article_section: Optional[str] = None


def _truncate(text: str, max_len: int = META_DESC_MAX_LEN) -> str:
    """Truncate and clean text for meta description."""
    if not text:
        return ""
    text = re.sub(r"\s+", " ", text).strip()
    if len(text) <= max_len:
        return text
    return text[: max_len - 3].rsplit(" ", 1)[0] + "..."


def _strip_html(text: str) -> str:
    """Remove HTML tags for plain-text meta."""
    if not text:
        return ""
    return re.sub(r"<[^>]+>", " ", text).strip()


def get_seo_for_page(
    page_type: str,
    *,
    title: Optional[str] = None,
    description: Optional[str] = None,
    image: Optional[str] = None,
    canonical_url: Optional[str] = None,
    category: Optional[str] = None,
    source: Optional[str] = None,
    published_iso: Optional[str] = None,
    article_id: Optional[str] = None,
    search_query: Optional[str] = None,
    page_number: Optional[int] = None,
    base_url: str = "",
    **kwargs: Any,
) -> SEOData:
    """
    Get professional SEO data for any page type. Use for all pages (home, article, category, search, etc.).

    Args:
        page_type: One of "home", "article", "category", "search", "not_found".
        title: Page or article title (used for article/detail pages).
        description: Raw description (will be truncated for meta).
        image: Absolute or relative image URL for og:image.
        canonical_url: Full canonical URL for the page.
        category: Category name (e.g. for category/list page or article section).
        source: Source name (e.g. for article author/source).
        published_iso: ISO 8601 date for article (og:article:published_time).
        article_id: Article ID (for canonical or future use).
        search_query: Search query (for search results page).
        page_number: Current page number (for paginated list).
        base_url: Base URL of the site (e.g. https://yoursite.com) for absolute canonical/image.

    Returns:
        SEOData with title, description, keywords, image, canonical_url, og_type, etc.
    """
    base = base_url.rstrip("/") if base_url else ""
    if image and not image.startswith(("http://", "https://", "//")):
        image = (base + image) if base else image

    if page_type == "home":
        seo_title = f"{SITE_NAME} - {SITE_TAGLINE}"
        seo_desc = (
            "Học tiếng Nhật qua tin tức với bản dịch tiếng Việt tự động. "
            "Đọc tin Nhật có dịch tiếng Việt miễn phí từ Mainichi và nguồn uy tín. "
            "Tin tức Mainichi tiếng Việt, học từ vựng và đọc hiểu tiếng Nhật qua bài báo mỗi ngày."
        )
        keywords = KEYWORDS_HOME
        return SEOData(
            title=seo_title,
            description=seo_desc,
            keywords=keywords,
            image=image or (base + DEFAULT_OG_IMAGE if base else None),
            canonical_url=canonical_url or (base + "/" if base else None),
            og_type="website",
        )

    if page_type == "article":
        display_title = (title or "Bài viết").strip()
        page_title = f"{display_title} | {SITE_NAME}"
        desc = description or ""
        desc = _strip_html(desc)
        seo_desc = _truncate(desc) if desc else _truncate(f"{display_title} - Đọc bài viết trên {SITE_NAME}.")
        keywords = KEYWORDS_HOME
        if category:
            keywords = f"{category}, {keywords}"
        return SEOData(
            title=page_title,
            description=seo_desc,
            keywords=keywords,
            image=image,
            canonical_url=canonical_url,
            og_type="article",
            article_published_time=published_iso,
            article_author=source,
            article_section=category,
        )

    if page_type == "category":
        cat_name = (category or "").strip()
        if not cat_name:
            # All categories view
            seo_title = f"Tất cả chuyên mục | {SITE_NAME}"
            seo_desc = (
                f"Xem tin tức tất cả chuyên mục: thế giới, kinh tế, công nghệ, khoa học, sức khỏe, thể thao, crypto, AI. "
                f"Tổng hợp và dịch sang tiếng Việt bởi {SITE_NAME}."
            )
            keywords = _keywords_all_categories()
        else:
            # Category-specific SEO with targeted keywords
            category_keywords = {
                "Tin chính": {
                    "title": f"Tin tức Nhật Bản - {cat_name} | {SITE_NAME}",
                    "desc": f"Tin tức Nhật Bản mới nhất về {cat_name}. Đọc tin thời sự Nhật có bản dịch tiếng Việt, học từ vựng và cải thiện kỹ năng đọc hiểu tiếng Nhật.",
                    "keywords": f"tin tức Nhật Bản, tin thời sự Nhật, {cat_name}, đọc tin Nhật có bản dịch, {KEYWORDS_HOME}"
                },
                "Thể thao": {
                    "title": f"Tin thể thao Nhật Bản | {SITE_NAME}",
                    "desc": f"Tin thể thao Nhật Bản mới nhất: bóng đá Nhật, sumo, bóng chày và các môn thể thao khác. Đọc tin thể thao có bản dịch tiếng Việt.",
                    "keywords": f"tin thể thao Nhật Bản, bóng đá Nhật, tin thể thao, {cat_name}, đọc tin Nhật có bản dịch, {KEYWORDS_HOME}"
                },
                "Giải trí": {
                    "title": f"Tin giải trí Nhật Bản | {SITE_NAME}",
                    "desc": f"Tin giải trí Nhật Bản mới nhất: showbiz Nhật, phim ảnh, âm nhạc. Đọc tin giải trí có bản dịch tiếng Việt, cập nhật xu hướng văn hóa Nhật.",
                    "keywords": f"tin giải trí Nhật Bản, showbiz Nhật, tin giải trí, {cat_name}, đọc tin Nhật có bản dịch, {KEYWORDS_HOME}"
                },
                "Chính luận": {
                    "title": f"Chính luận & Bình luận Nhật Bản | {SITE_NAME}",
                    "desc": f"Các bài xã luận và bình luận Nhật Bản về các vấn đề quan trọng. Đọc chính luận có bản dịch tiếng Việt để hiểu sâu hơn về xã hội Nhật.",
                    "keywords": f"bình luận Nhật Bản, xã luận Nhật, chính luận, {cat_name}, đọc tin Nhật có bản dịch, {KEYWORDS_HOME}"
                }
            }
            
            cat_seo = category_keywords.get(cat_name, {
                "title": f"{cat_name} | {SITE_NAME}",
                "desc": f"Tin tức mới nhất về {cat_name}. Tổng hợp và dịch sang tiếng Việt bởi {SITE_NAME}.",
                "keywords": f"{cat_name}, tin tức, {KEYWORDS_HOME}"
            })
            
            seo_title = cat_seo["title"]
            seo_desc = cat_seo["desc"]
            keywords = cat_seo["keywords"]
        return SEOData(
            title=seo_title,
            description=_truncate(seo_desc) if not cat_name else seo_desc,
            keywords=keywords,
            image=image or (base + DEFAULT_OG_IMAGE if base else None),
            canonical_url=canonical_url,
            og_type="website",
        )

    if page_type == "search":
        query = (search_query or "").strip() or "Tìm kiếm"
        seo_title = f"Tìm kiếm: {query} | {SITE_NAME}"
        seo_desc = f"Kết quả tìm kiếm tin tức cho \"{query}\" trên {SITE_NAME}. Tổng hợp tin đa nguồn, dịch sang tiếng Việt."
        keywords = f"tìm kiếm, {query}, {KEYWORDS_HOME}"
        return SEOData(
            title=seo_title,
            description=_truncate(seo_desc),
            keywords=keywords,
            image=image or (base + DEFAULT_OG_IMAGE if base else None),
            canonical_url=canonical_url,
            og_type="website",
        )

    if page_type == "not_found":
        seo_title = f"Không tìm thấy trang | {SITE_NAME}"
        seo_desc = f"Trang bạn tìm kiếm không tồn tại. Quay lại {SITE_NAME} để xem tin tức mới nhất."
        return SEOData(
            title=seo_title,
            description=seo_desc,
            keywords=KEYWORDS_HOME,
            image=image or (base + DEFAULT_OG_IMAGE if base else None),
            canonical_url=canonical_url,
            og_type="website",
        )

    # Generic fallback
    seo_title = title or SITE_NAME
    if seo_title != SITE_NAME and not seo_title.endswith(SITE_NAME):
        seo_title = f"{seo_title} | {SITE_NAME}"
    seo_desc = _truncate(description or SITE_TAGLINE)
    return SEOData(
        title=seo_title,
        description=seo_desc,
        keywords=KEYWORDS_HOME,
        image=image,
        canonical_url=canonical_url,
        og_type=kwargs.get("og_type", "website"),
    )


def seo_to_dict(seo: SEOData) -> dict:
    """Convert SEOData to a flat dict for API/JSON."""
    d = {
        "title": seo.title,
        "description": seo.description,
        "keywords": seo.keywords,
        "image": seo.image,
        "canonical_url": seo.canonical_url,
        "og_type": seo.og_type,
    }
    if seo.article_published_time:
        d["article_published_time"] = seo.article_published_time
    if seo.article_author:
        d["article_author"] = seo.article_author
    if seo.article_section:
        d["article_section"] = seo.article_section
    return d

File: app/static/js/seo.js
/**
 * SEO module - used for all pages. Sets document title, meta description,
 * keywords, Open Graph and Twitter Card tags for professional, consistent SEO.
 */
(function (global) {
    'use strict';

    var SITE_NAME = 'Sugoi News';
    var META_DESC_MAX = 160;

    function escapeAttr(str) {
        if (str == null || str === '') return '';
        var div = document.createElement('div');
        div.textContent = str;
        return div.innerHTML.replace(/"/g, '&quot;');
    }

    function setOrCreateMeta(attr, value, isProperty) {
        if (!value) return;
        var name = isProperty ? 'property' : 'name';
        var selector = isProperty ? 'meta[property="' + escapeAttr(attr) + '"]' : 'meta[name="' + escapeAttr(attr) + '"]';
        var el = document.querySelector(selector);
        if (!el) {
            el = document.createElement('meta');
            el.setAttribute(name, attr);
            document.head.appendChild(el);
        }
        el.setAttribute('content', value);
    }

    function setOrCreateLink(rel, href) {
        if (!href) return;
        var el = document.querySelector('link[rel="' + rel + '"]');
        if (!el) {
            el = document.createElement('link');
            el.setAttribute('rel', rel);
            document.head.appendChild(el);
        }
        el.setAttribute('href', href);
    }

    /**
     * Set page SEO for all pages. Call on every page (home, article, category, search).
     * @param {Object} options - SEO options (from API /api/seo or built manually)
     * @param {string} [options.title] - Page title (document.title)
     * @param {string} [options.description] - Meta description (truncate to ~160 chars)
     * @param {string} [options.keywords] - Meta keywords
     * @param {string} [options.image] - Absolute image URL for og:image / twitter:image
     * @param {string} [options.canonical_url] - Canonical URL
     * @param {string} [options.og_type] - og:type (website | article)
     * @param {string} [options.article_published_time] - ISO date for article
     * @param {string} [options.article_author] - Author/source for article
     * @param {string} [options.article_section] - Section/category for article
     */
    function setPageSEO(options) {
        if (!options || typeof options !== 'object') return;
        var title = (options.title || '').trim();
        var description = (options.description || '').trim();
        if (description.length > META_DESC_MAX) {
            description = description.slice(0, META_DESC_MAX - 3).replace(/\s+\S*$/, '') + '...';
        }
        var keywords = (options.keywords || '').trim();
        var image = (options.image || '').trim();
        var canonicalUrl = (options.canonical_url || '').trim();
        var ogType = (options.og_type || 'website').toLowerCase();
        var baseUrl = (options.base_url || window.location.origin || '').replace(/\/$/, '');
        if (image && !/^https?:\/\//i.test(image)) {
            image = baseUrl + (image.indexOf('/') === 0 ? image : '/' + image);
        }
        // If no canonical URL provided, use current URL (without hash/fragment)
        if (!canonicalUrl) {
            canonicalUrl = window.location.origin + window.location.pathname + window.location.search;
        } else if (!/^https?:\/\//i.test(canonicalUrl)) {
            canonicalUrl = baseUrl + (canonicalUrl.indexOf('/') === 0 ? canonicalUrl : '/' + canonicalUrl);
        }
        var currentUrl = window.location.href;

        if (title) document.title = title;

        setOrCreateMeta('description', description, false);
        if (keywords) setOrCreateMeta('keywords', keywords, false);

        setOrCreateMeta('og:title', title, true);
        setOrCreateMeta('og:description', description, true);
        setOrCreateMeta('og:type', ogType, true);
        setOrCreateMeta('og:site_name', SITE_NAME, true);
        setOrCreateMeta('og:url', currentUrl, true);
        if (image) {
            setOrCreateMeta('og:image', image, true);
            setOrCreateMeta('og:image:width', '1200', true);
            setOrCreateMeta('og:image:height', '630', true);
        }

        setOrCreateMeta('twitter:card', 'summary_large_image', false);
        setOrCreateMeta('twitter:title', title, false);
        setOrCreateMeta('twitter:description', description, false);
        if (image) setOrCreateMeta('twitter:image', image, false);

        if (canonicalUrl) setOrCreateLink('canonical', canonicalUrl);

        if (ogType === 'article') {
            if (options.article_published_time) setOrCreateMeta('article:published_time', options.article_published_time, true);
            if (options.article_author) setOrCreateMeta('article:author', options.article_author, true);
            if (options.article_section) setOrCreateMeta('article:section', options.article_section, true);
        }
    }

    /**
     * Fallback SEO for category pages (when /api/seo fails). Matches app.js categoryNames.
     */
    function getCategorySEOFallback(categoryName) {
        var map = {
            'Tin chính': {
                title: 'Tin tức Nhật Bản - Tin chính',
                description: 'Tin tức mới nhất về các sự kiện quan trọng tại Nhật Bản. Đọc tin chính có bản dịch tiếng Việt, học từ vựng và cải thiện kỹ năng đọc hiểu tiếng Nhật.',
                keywords: 'tin tức Nhật Bản, tin thời sự Nhật, học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, học tiếng Nhật qua đọc tin tức, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, học từ vựng tiếng Nhật qua tin tức, đọc hiểu tiếng Nhật qua bài báo, Sugoi News'
            },
            'Thể thao': {
                title: 'Tin thể thao Nhật Bản',
                description: 'Tin tức thể thao mới nhất từ Nhật Bản: bóng đá, sumo, bóng chày và các môn thể thao khác. Đọc tin thể thao có bản dịch tiếng Việt.',
                keywords: 'tin thể thao Nhật Bản, bóng đá Nhật, học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, đọc hiểu tiếng Nhật qua bài báo, Sugoi News'
            },
            'Giải trí': {
                title: 'Tin giải trí Nhật Bản',
                description: 'Tin tức giải trí, showbiz, phim ảnh, âm nhạc Nhật Bản. Đọc tin giải trí có bản dịch tiếng Việt, cập nhật xu hướng văn hóa Nhật.',
                keywords: 'tin giải trí Nhật Bản, showbiz Nhật, học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, đọc hiểu tiếng Nhật qua bài báo, Sugoi News'
            },
            'Chính luận': {
                title: 'Chính luận & Bình luận Nhật Bản',
                description: 'Các bài xã luận, bình luận và phân tích về các vấn đề quan trọng tại Nhật Bản. Đọc chính luận có bản dịch tiếng Việt để hiểu sâu hơn về xã hội Nhật.',
                keywords: 'bình luận Nhật Bản, xã luận Nhật, học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, đọc hiểu tiếng Nhật qua bài báo, Sugoi News'
            }
        };
        var info = map[categoryName];
        if (info) return info;
        return {
            title: categoryName,
            description: 'Tin tức mới nhất về ' + categoryName + '. Tổng hợp và dịch sang tiếng Việt bởi Sugoi News. Học tiếng Nhật qua tin tức.',
            keywords: categoryName + ', học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, đọc hiểu tiếng Nhật qua bài báo, Sugoi News'
        };
    }

    /**
     * Fetch SEO data from API and apply. Use for home, category, search, or article (with article_id).
     * @param {Object} params - { page: 'home'|'article'|'category'|'search'|'not_found', article_id?, category?, search? }
     * @returns {Promise<Object>} - Fetched SEO object (or default/fallback SEO on error)
     */
    function fetchAndSetSEO(params) {
        params = params || {};
        var qs = new URLSearchParams();
        qs.set('page', params.page || 'home');
        if (params.article_id) qs.set('article_id', params.article_id);
        if (params.category) qs.set('category', params.category);
        if (params.search) qs.set('search', params.search);
        var baseUrl = window.location.origin.replace(/\/$/, '');
        return fetch('/api/seo?' + qs.toString()).then(function (r) {
            return r.json();
        }).then(function (data) {
            data.base_url = baseUrl;
            setPageSEO(data);
            return data;
        }).catch(function () {
            var page = params.page || 'home';
            var opts = { base_url: baseUrl, og_type: 'website' };
            if (page === 'category' && params.category) {
                var catSEO = getCategorySEOFallback(params.category);
                opts.title = (catSEO.title || params.category) + ' | ' + SITE_NAME;
                opts.description = (catSEO.description && catSEO.description.length > META_DESC_MAX)
                    ? catSEO.description.slice(0, META_DESC_MAX - 3).replace(/\s+\S*$/, '') + '...'
                    : (catSEO.description || '');
                opts.keywords = catSEO.keywords || '';
                opts.canonical_url = baseUrl + '/?category=' + encodeURIComponent(params.category);
            } else if (page === 'search' && params.search) {
                opts.title = 'Tìm kiếm: ' + params.search + ' | ' + SITE_NAME;
                opts.description = 'Kết quả tìm kiếm tin tức cho "' + params.search + '" trên Sugoi News. Tổng hợp tin đa nguồn, dịch sang tiếng Việt. Học tiếng Nhật qua tin tức.';
                opts.keywords = params.search + ', tìm kiếm tin tức Nhật Bản, Sugoi News';
                opts.canonical_url = baseUrl + '/?search=' + encodeURIComponent(params.search);
            } else {
                opts.title = SITE_NAME + ' - Học tiếng Nhật qua tin tức';
                opts.description = 'Sugoi News - Học tiếng Nhật qua tin tức. Đọc tin Nhật có dịch tiếng Việt miễn phí từ Mainichi. Tin tức Mainichi tiếng Việt, học từ vựng và đọc hiểu tiếng Nhật qua bài báo mỗi ngày.';
                opts.keywords = 'học tiếng Nhật qua tin tức, đọc tin Nhật có bản dịch, tin tức Nhật Bản tiếng Việt, học tiếng Nhật qua đọc tin tức, đọc tin Nhật có dịch tiếng Việt miễn phí, tin tức Mainichi tiếng Việt, học từ vựng tiếng Nhật qua tin tức, đọc hiểu tiếng Nhật qua bài báo, Sugoi News';
            }
            setPageSEO(opts);
            return null;
        });
    }

    global.setPageSEO = setPageSEO;
    global.fetchAndSetSEO = fetchAndSetSEO;
})(typeof window !== 'undefined' ? window : this);
